test/Microsoft.OpenApi.Tests/Services/OpenApiComponentsTests.cs

```diff
@@ -0,0 +1,610 @@
+                                }
+                            }
+                        },
+                        ["requestBody2"] = new OpenApiRequestBody
+                        {
+                            Description = "description",
+                            Required = true,
+                            Content =
+                            {
+                                ["application/json"] = new OpenApiMediaType
+                                {
+                                    Schema = new OpenApiSchema
+                                    {
+                                        Type = "string"
+                                    }
+                                }
+                            }
+                        }
+                    }
+                },
+                new List<OpenApiDifference>
+                {
+                    new OpenApiDifference
+                    {
+                        Pointer = "#/requestBodies/requestBody2",
+                        OpenApiDifferenceOperation = OpenApiDifferenceOperation.Add,
+                        OpenApiComparedElementType = typeof(KeyValuePair<string, OpenApiRequestBody>)
+                    },
+                    new OpenApiDifference
+                    {
+                        Pointer =
+                            "#/schemas/schemaObject2",
+                        OpenApiDifferenceOperation = OpenApiDifferenceOperation.Add,
+                        OpenApiComparedElementType = typeof(KeyValuePair<string, OpenApiSchema>)
+                    }
+                }
+            };
+        }
+
+        [Theory]
+        [MemberData(nameof(GetTestCasesForOpenApiComponentsComparerShouldSucceed))]
+        public void OpenApiComponentsComparerShouldSucceed(
+            string testCaseName,
+            OpenApiComponents source,
+            OpenApiComponents target,
+            List<OpenApiDifference> expectedDifferences)
+        {
+            _output.WriteLine(testCaseName);
+
+            var comparisonContext = new ComparisonContext(new OpenApiComparerFactory(), _sourceDocument,
+                _targetDocument);
+            var comparer = new OpenApiComponentsComparer();
+            comparer.Compare(source, target, comparisonContext);
+
+            var differences = comparisonContext.OpenApiDifferences.ToList();
+
+            differences.Count().ShouldBeEquivalentTo(expectedDifferences.Count);
+
+            var notExpectedDifferences = differences.Where(
+                actualDiff => !expectedDifferences.Any(
+                    expectedDiff => expectedDiff.Pointer == actualDiff.Pointer
+                        && expectedDiff.OpenApiComparedElementType == actualDiff.OpenApiComparedElementType
+                        && expectedDiff.OpenApiDifferenceOperation ==
+                            actualDiff.OpenApiDifferenceOperation)).ToList();
+
+            notExpectedDifferences.Count.ShouldBeEquivalentTo(0);
```

Comment:
Can't we just do

differences.ShouldBeEquivalentTo(expectedDifferences) ?

Reply:
Same comment with other test files.

---
In reply to: [220743939](https://github.com/Microsoft/OpenAPI.NET/pull/316#discussion_r220743939) [](ancestors = 220743939)

Reply:
So i can't do that, as in the difference result we have SourceValue and TargetValue of type object and equivalent fails when i use the actual type e.g. Server in expected result.

---
In reply to: [220743996](https://github.com/Microsoft/OpenAPI.NET/pull/316#discussion_r220743996) [](ancestors = 220743996,220743939)

Reply:
ShouldBeEquivalentTo should be able to to do a deep comparison into any object type. In your expectedDifferences, you can assign the correct sourceValue and targetValue and FluentAssertion should be able to compare it with the differences you got.

In the case there's a bug in FluentAssertion library and the above doesn't work, you can simply exclude certain path from comparison in fluent assertion. [https://stackoverflow.com/questions/22142576/how-to-use-exclude-in-fluentassertions-for-property-in-collection](https://stackoverflow.com/questions/22142576/how-to-use-exclude-in-fluentassertions-for-property-in-collection)

This will ensure this test has good coverage even when we update OpenApiDifference in the future.

---
In reply to: [221082998](https://github.com/Microsoft/OpenAPI.NET/pull/316#discussion_r221082998) [](ancestors = 221082998,220743996,220743939)

Reply:
interesting it worked, i remember it not working somewhere will fix all test.

---
In reply to: [221094506](https://github.com/Microsoft/OpenAPI.NET/pull/316#discussion_r221094506) [](ancestors = 221094506,221082998,220743996,220743939)

Reply:
It actually does compare, right? Did you try testing it by intentionally making it fail? Just want to make sure it's not just giving out an automatic pass without comparing.

---
In reply to: [221702898](https://github.com/Microsoft/OpenAPI.NET/pull/316#discussion_r221702898) [](ancestors = 221702898,221094506,221082998,220743996,220743939)

Reply:
yes tested it, it does compare it and not false positive.

---
In reply to: [221749561](https://github.com/Microsoft/OpenAPI.NET/pull/316#discussion_r221749561) [](ancestors = 221749561,221702898,221094506,221082998,220743996,220743939)